src/RelationalBehavior.php

```diff
@@ -113,40 +126,71 @@ public function saveRelations($event)
                 list($junctionTable, $primaryModelColumn, $relatedModelColumn) = $this->parseQuery($activeQuery);
-
                 $junctionRows = [];
                 $relationPks = ArrayHelper::getColumn($relationRecords, array_keys($activeQuery->link)[0], false);
                 $passedRecords = count($relationPks);
-                $relationPks = array_filter($relationPks);
+
+                if ($this->sortSettings[$relationName] &&
+                    isset($this->sortSettings[$relationName]['sortColumn']) &&
+                    isset($this->relationalRawData[$relationName])
+                ) {
+                    $relationPks = $this->relationalRawData[$relationName];
+                } else {
+                    $relationPks = array_filter($relationPks);
+                }
+
                 $savedRecords = count($relationPks);
                 if ($passedRecords != $savedRecords) {
-                    throw new RuntimeException('All relation records must be saved');
+                    throw new RuntimeException('All relation records must be saved. There are incorrect PKs.');
                 }
                 foreach ($relationPks as $relationPk) {
-                    $junctionRows[] = [$model->primaryKey, $relationPk];
+
+                    $position++;
+
+                    if (isset($this->sortSettings[$relationName]['sortColumn'])) {
+                        $junctionRows[] = [$model->primaryKey, $relationPk, $position];
+                    } else {
+                        $junctionRows[] = [$model->primaryKey, $relationPk];
+                    }
                 }
 
-                $model->getDb()->transaction(function() use ($junctionTable, $primaryModelColumn, $relatedModelColumn,
-                    $junctionRows, $model
+                $model->getDb()->transaction(function () use (
+                    $junctionTable,
+                    $primaryModelColumn,
+                    $relatedModelColumn,
+                    $junctionRows,
+                    $model,
+                    $relationName
                 ) {
                     $db = $model->getDb();
-                    $db->createCommand()->delete($junctionTable, [$primaryModelColumn => $model->primaryKey])->execute();
+                    $db->createCommand()->delete($junctionTable,
+                        [$primaryModelColumn => $model->primaryKey])->execute();
                     if (!empty($junctionRows)) {
-                        $db->createCommand()->batchInsert($junctionTable, [$primaryModelColumn, $relatedModelColumn],
-                            $junctionRows)->execute();
+
+                        if (isset($this->sortSettings[$relationName]['sortColumn'])) {
+                            $db->createCommand()->batchInsert($junctionTable,
+                                [$primaryModelColumn, $relatedModelColumn, $this->sortSettings[$relationName]['sortColumn']],
```

Comment:
I would extract this array and modify in condition. Duplicating batchInsert calls breaks DRY principle